Maple2.Server.Game/Model/Field/Actor/ActorState/AiState.cs

```diff
@@ -0,0 +1,452 @@
+            return true;
+        }
+
+        return false;
+    }
+
+    private void Process(Entry entry) {
+        if (entry is Node) {
+            actor.AppendDebugMessage($"> Node: {entry.Name}\n");
+
+            ProcessNode((dynamic) entry);
+
+            return;
+        }
+
+    }
+
+    private void ProcessNode(TraceNode node) {
+        if (!actor.Field.TryGetActor(actor.TargetId, out IActor? target)) {
+            return;
+        }
+
+        actor.Navigation.PathTo(target.Position);
+    }
+
+    private void ProcessNode(SkillNode node) {
+        //new SkillRecord()
+        //actor.Field?.Broadcast(SkillPacket.Use());
+        int skillId = 0;
+        short skillLevel = 0;
+
+        NpcMetadataSkill[] skills = actor.Value.Metadata.Property.Skills;
+
+        // idx starts at 1
+        if (node.Idx > skills.Length) {
+            actor.AppendDebugMessage($"{actor.Value.Metadata.Name}[{actor.Value.Id}]\n");
+            actor.AppendDebugMessage($"{ AiMetadata!.Name}\n");
+            actor.AppendDebugMessage($"Invalid Skill Idx {node.Idx}\n");
+
+            return;
+        }
+
+        skillId = skills[node.Idx - 1].Id;
+        skillLevel = (short)skills[node.Idx - 1].Level;
+
+        SkillRecord? cast = actor.CastSkill(skillId, skillLevel);
+
+        if (cast is null) {
+            actor.AppendDebugMessage($"Failed cast {skillId}[{skillLevel}]\n");
+
+            return;
+        }
+    }
+
+    private void ProcessNode(TeleportNode node) {
+        actor.Position = node.Pos;
+    }
+
+    private void ProcessNode(StandbyNode node) {
+        actor.Navigation.PathTo(actor.Position); // stop moving
+    }
+
+    private void ProcessNode(SetDataNode node) {
+
+    }
+
+    private void ProcessNode(TargetNode node) {
+
+    }
+
+    private void ProcessNode(SayNode node) {
+        actor.Field.Broadcast(CinematicPacket.BalloonTalk(true, actor.ObjectId, node.Message, node.DurationTick, node.DelayTick));
+
+        actor.AppendDebugMessage($"Say '{node.Message}'");
+    }
+
+    private void ProcessNode(SetValueNode node) {
+
+    }
+
+    private void ProcessNode(ConditionsNode node) {
+        Condition? passed = null;
+
+        foreach (Condition condition in node.Conditions) {
+            if (ProcessCondition((dynamic)condition)) {
+                passed = condition;
+
+                break;
+            }
+        }
+
+        if (passed is null) {
+            actor.AppendDebugMessage("Failed conditions\n");
+
+            return;
+        }
+
+        actor.AppendDebugMessage($"+ Condition: '{passed.Name}'\n");
+
+        Push(passed);
+    }
+
+    private void ProcessNode(JumpNode node) {
+
+    }
+
+    private void ProcessNode(SelectNode node) {
+        int totalWeight = 0;
+
+        foreach (int weight in node.Prob) {
+            totalWeight += weight;
+        }
+
+        int roll = Random.Shared.Next(0, totalWeight);
+        int selected = -1;
+
+        for (int i = 0; i < node.Prob.Length; ++i) {
```

Comment:
could weightedset help here?

Reply:
I'm not sure it'd make too much of a difference. The amount of scaffolding code in other areas that use that looks roughly the same, and it comes with the added cost of some allocations that are otherwise avoided here.

Reply:
I'd rather we use weightedsets than reinvent the wheel. It was literally meant for things like this. Refer to ItemDropManager if you need to.